### ISSUE TITLE
Path bug in Windows

### ISSUE THREAD

**Issue by just1a-person** [foambubble/foam#488]
- Foam version: v0.9.1
- Platform: Windows 
- Issue occur on the [foam template](https://github.com/foambubble/foam-template) repo: Not sure

**Summary**

![Windows wiki-link bug](https://user-images.githubusercontent.com/75578443/107957065-131ff000-6fc6-11eb-8d9a-82743246efb2.png)

The Windows path is very weird. It's not at all portable because it is not a relative path.

**Steps to reproduce**
1.  Add a link to todo.md by [[todo]]
2.  Ctrl + S

**Additional information**

Feel free to attach any of the following that might help with debugging the issue:
- screenshots
- a zip with a minimal repo to reproduce the issue
- the Foam log in VsCode (see [instructions](https://github.com/foambubble/foam/blob/master/docs/foam-logging-in-vscode.md))

**Comment by itsme-alan** [comment]
I can still click on it and it opens the right file. But only on my main machine because it is not relative.

**Comment by itsme-alan** [comment]
Looks like there is a whole lot of bugs with path handling in Windows. same files are duplicated in the graph. Looking at the logs, Foam seems to think that there is two files with the same name.
```
[info - 8:25:47 pm] Warning: Slug CSharp matches 2 documents. Picking one.
[info - 8:25:47 pm] Updated: /g:/MyFoam/Programming/ThingsThatIAlwaysHaveToLookUp/
```

In reality there is only one CSharp.md.

First class Windows support is something that I expect before 1.0

**Comment by riccardoferretti** [comment]
#467 will address the duplication.
once it's merged let me know if the issue about the link definitions is still present, I am a bit surprised to see that

**Comment by kenotron** [comment]
Pretty sure this is due to the difference in casing of the drive letter. Still happens for me with the 10.2 release

**Comment by kenotron** [comment]
Pretty sure the bug is inside the uri.ts: `computeRelativePath()`

**Comment by joshkreud** [comment]
Heyo, 

I'm facing the same issue on windows. 
The issue started within the last 1 or 2 releases. 

As already mentioned, `computeRelativePath()` seems to be a good starting point, so I did a bit of testing: 

Apparently aligning the case of both paths fixes the `posix.relative()` call. 
```js
export const computeRelativePath = (source: URI, target: URI): string => {
  const sourceDir = posix.dirname(source.path);
  const ucaseSource = sourceDir.toUpperCase();
  const ucaseTarget = target.path.toUpperCase();
  const relativePath = posix.relative(ucaseSource, ucaseTarget);
  return relativePath;
};
```
But when i do this, saving a file does not add all the Link references, but instead just Toggles though them (each save shows another link target) 

So back to the beginning. 
 `computeRelativePath()`  is used in `createMarkdownReferences()`. 

There we get the `noteUri` as an argument. This is the source path with the capital drive letter.
And the target path, with the lower case drive letter comes from `FoamWorkspace.find()`

I'll investigate a bit further. 
Js is sometimes still a bit unfriendly to me... So to be continued.

**Comment by riccardoferretti** [comment]
Thanks for looking into this, I don't have a Win machine so it's pretty hard for me to debug this issue. Thoughts:
- IIUC the issue is that in one case the drive letter is upper case, whereas in the other it's not? Another issue (#494) also seems related to filename cases, linking.
- `computeRelativePath` should probably be called `computePosixRelativePath` - currently it's only used in one place (where that's what we want), but I can see a source of confusion here
- we can add some test cases for this function, so that they will run in CI which includes a Win environment

> But when I do this, saving a file does not add all the Link references, but instead just Toggles though them (each save shows another link target)

Not sure what you mean here?

> Js is sometimes still a bit unfriendly to me

It's not about you, it's about js ;)

**Comment by joshkreud** [comment]
Hey

> IIUC the issue is that in one case the drive letter is upper case, whereas in the other it's not? Another issue (#494) also seems related to filename cases, linking.

Exactly. `posix.relative()` fails, if they differ in case. 

> But when I do this, saving a file does not add all the Link references, but instead just Toggles though them (each save shows another link target)
Not sure what you mean here?

Okay I'll try to be a bit more specific: 
Windows actually doesn't care about the case in a Path. So back in my VBA days, used to handle all paths simply as Upper Case. 
I know ... hacky, but well ... windows .... 
When I try this within foam I guess it doesn't pass some check further down the road. 
 
So lets say I have 2 Links in a file. 
This should produce 2 link entries at the bottom of the file. 
But when I apply the upper case workaround, it doesn't add both. 
It alternates between the two whenever the file gets saved. (It only shows one entry) 


So basically like this: 
![image](https://user-images.githubusercontent.com/10895412/109510562-7455c200-7aa2-11eb-9c23-6ff0afd58152.png)
And another Save produces this: 
![image](https://user-images.githubusercontent.com/10895412/109510601-80418400-7aa2-11eb-940e-8c776c441d12.png)
and another save goes back to  the first image and so on. 


But this would just be a symptomatic fix, not the root cause. 
I'll dig a bit deeper into Foam, lets see what I can find.

**Comment by kenotron** [comment]
I don't think the right way is to completely disregard the casing by forcing upper case on Windows. An idea is to write a "normalizeDriveLetter" that would surgically normalize that portion of the path if it's on Windows.

**Comment by kenotron** [comment]
So I created #507 to address this issue, can you please take a look, @riccardoferretti?

**Comment by itsme-alan** [comment]
Is this fixed now? I had decided to start using Foam again after they fix this issue.

**Comment by riccardoferretti** [comment]
This should now be fixed, can you please confirm?

**Comment by itsme-alan** [comment]
Yes. It is working.

![Links show up correctly in output, screenshot](https://user-images.githubusercontent.com/75578443/124244557-b1a78a80-db3c-11eb-97e5-30fa485c3886.png)

**Comment by joshkreud** [comment]
For me the error is still appearing on Windows. 
But as an easy workaround I just have my Foam in the WSL. Linux is just better to work with.

**Comment by itsme-alan** [comment]
Are you updated to the latest version of Foam?

**Comment by joshkreud** [comment]
I didn't have foam installed on the Windows's where I tested it, so I suppose It installed the newest version.

**Comment by itsme-alan** [comment]
That's odd. It works for me (as seen in the screenshot)

I am going to keep this issue open anyway.

**Comment by zcysxy** [comment]
This bug is finally fixed on my side in the newest version 🎉. Thanks @riccardoferretti 🙌 . Does it get fixed on your end @joshkreud?

**Comment by riccardoferretti** [comment]
Glad to hear that  :)

Given that we had to make a change to a pretty core class, feel free to play around and let me know if you discover any issue. (I am optimistic, but cautiously so ;) ).

All of this doesn't change my feelings towards (against) Windows paths hehe
